boot/bootutil/include/bootutil/bootutil_public.h

```diff
@@ -0,0 +1,184 @@
+
+/*
+ * NOTE: leave BOOT_FLAG_SET equal to one, this is written to flash!
+ */
+#define BOOT_FLAG_SET       1
+#define BOOT_FLAG_BAD       2
+#define BOOT_FLAG_UNSET     3
+#define BOOT_FLAG_ANY       4  /* NOTE: control only, not dependent on sector */
+
+#define BOOT_MAGIC_SZ (sizeof boot_img_magic)
+
+struct boot_swap_state {
+    uint8_t magic;      /* One of the BOOT_MAGIC_[...] values. */
+    uint8_t swap_type;  /* One of the BOOT_SWAP_TYPE_[...] values. */
+    uint8_t copy_done;  /* One of the BOOT_FLAG_[...] values. */
+    uint8_t image_ok;   /* One of the BOOT_FLAG_[...] values. */
+    uint8_t image_num;  /* Boot status belongs to this image */
+};
+
+/**
+ * @brief Determines the action, if any, that mcuboot will take on a image pair.
+ *
+ * @param image_index Image pair index.
+ *
+ * @return a BOOT_SWAP_TYPE_[...] constant on success, negative errno code on
+ * fail.
+ */
+int boot_swap_type_multi(int image_index);
+
+/**
+ * @brief Determines the action, if any, that mcuboot will take.
+ *
+ * Works the same as a boot_swap_type_multi(0) call;
+ *
+ * @return a BOOT_SWAP_TYPE_[...] constant on success, negative errno code on
+ * fail.
+ */
+int boot_swap_type(void);
+
+/**
+ * Marks the image in the secondary slot as pending. On the next reboot,
+ * the system will perform a one-time boot of the the secondary slot image.
+ *
+ * @param permanent Whether the image should be used permanently or
+ *                  only tested once:
+ *                    0=run image once, then confirm or revert.
+ *                    1=run image forever.
+ *
+ * @return 0 on success; nonzero on failure.
+ */
+int boot_set_pending(int permanent);
+
+/**
+ * @brief Marks the image in the primary slot as confirmed.
+ *
+ * The system will continue booting into the image in the primary slot until
+ * told to boot from a different slot.
+ *
+ * @return 0 on success; nonzero on failure.
+ */
+int boot_set_confirmed(void);
+
+/**
+ * @brief Get offset of the swap info field in the image trailer.
+ *
+ * @param fap Flash are for which offset is determined.
+ *
+ * @retval offset of the swap info field.
+ */
+uint32_t boot_swap_info_off(const struct flash_area *fap);
+
+/**
+ * @brief Get value of image-ok flag of the image.
+ *
+ * If called from chin-loaded image the image-ok flag value can be used to check
+ * whether application itself is already confirmed.
+ *
+ * @param fap Flash area of the image.
+ * @param image_ok[out] image-ok value.
+ *
+ * @return 0 on success; nonzero on failure.
+ */
+int boot_read_image_ok(const struct flash_area *fap, uint8_t *image_ok);
+
+/**
+ * @brief Read the image swap state
+ *
+ * @param flash_area_id Id of flash parttition from which trailer will be read.
+ * @param state Struture for holding swap state.
+ *
+ * @return 0 on success, nonzero errno code on fail.
+ */
+int
+boot_read_swap_state_by_id(int flash_area_id, struct boot_swap_state *state);
+
```

Comment:
Considering the OTA case, the tfm firmware update partition needs to get the image version information. This can be achieved by reading the shared memory between the bl2 and the tfm, but I think adding such an interface like boot_query_image_info() in bl2 is more directly and user friendly to get the image version or image hash for the tfm partitions. Also some users download and writes the image directly in the nonsecure side, in this case, the nonsecure does not have access to the shared memory of tfm and bl2, so it is necessary to add such image query API.

Reply:
Also, I think, the public API like boot_write_secondary_image or simply flash_area_write should be added to the public library. These APIs are used by the firmware update service in TF-M or the nonsecure firmware update application. The firmware update service, as well as the nonsecure firmware update application, should not all the flash_area_write in the mcuboot src, instead, it should call the public library directly which should cover all the functionality for the firmware update.
If this PR is just to extract the currently existing public APIs, I can create a new PR to add the 'query' and 'write' functions to the public library.

Reply:
Yes: this PR is mainly for extract the current existing public APIs.
I'm not sure why do you mean by `flash_area_write` here. flash_area_write() is on of the porting functions.
https://github.com/mcu-tools/mcuboot/blob/master/docs/PORTING.md

`boot_write_secondary_image()` Is this function for collecting the image to the second image bank? Such functionalities are supposed (and are) implemented by the application. See no point to implement this here.

Reply:
Make sense. It seems the flash_area_write and other flash operations are implemented in TF-M, not in the mcuboot repo.